packages/test/test-end-to-end-tests/src/test/idCompressor.spec.ts

```diff
@@ -0,0 +1,326 @@
+	return (dds as any).runtime.idCompressor as IIdCompressor;
+}
+
+describeNoCompat("Runtime IdCompressor", (getTestObjectProvider) => {
+	let provider: ITestObjectProvider;
+	beforeEach(() => {
+		provider = getTestObjectProvider();
+	});
+
+	let sharedMap1: SharedMap;
+	let sharedCell1: SharedCell;
+	let sharedMap2: SharedMap;
+	let sharedMap3: SharedMap;
+
+	beforeEach(async () => {
+		const container1 = await provider.makeTestContainer(testContainerConfig);
+		const dataObject1 = await requestFluidObject<ITestFluidObject>(container1, "default");
+		sharedMap1 = await dataObject1.getSharedObject<SharedMap>(mapId);
+		sharedCell1 = await dataObject1.getSharedObject<SharedCell>(cellId);
+
+		const container2 = await provider.loadTestContainer(testContainerConfig);
+		const dataObject2 = await requestFluidObject<ITestFluidObject>(container2, "default");
+		sharedMap2 = await dataObject2.getSharedObject<SharedMap>(mapId);
+
+		const container3 = await provider.loadTestContainer(testContainerConfig);
+		const dataObject3 = await requestFluidObject<ITestFluidObject>(container3, "default");
+		sharedMap3 = await dataObject3.getSharedObject<SharedMap>(mapId);
+
+		sharedMap1.set("testKey1", "testValue");
+
+		await provider.ensureSynchronized();
+	});
+
+	it("has no compressor if not enabled", async () => {
+		provider.reset();
+		const config: ITestContainerConfig = {
+			registry: [
+				[mapId, SharedMap.getFactory()],
+				[cellId, SharedCell.getFactory()],
+			],
+			fluidDataObjectType: DataObjectFactoryType.Test,
+		};
+		const container1 = await provider.makeTestContainer(config);
+		const dataObject1 = await requestFluidObject<ITestFluidObject>(container1, "default");
+		const map = await dataObject1.getSharedObject<SharedMap>(mapId);
+
+		assert(getIdCompressor(map) === undefined);
+	});
+
+	it("can normalize session space IDs to op space", async () => {
+		assert(getIdCompressor(sharedMap1) !== undefined);
+		assert(getIdCompressor(sharedMap2) !== undefined);
+		assert(getIdCompressor(sharedMap3) !== undefined);
+
+		const compressedIds: Map<SessionSpaceCompressedId, SharedMap> = new Map();
+		for (let i = 0; i < 1024; i++) {
+			compressedIds.set(getIdCompressor(sharedMap1).generateCompressedId(), sharedMap1);
+			compressedIds.set(getIdCompressor(sharedMap2).generateCompressedId(), sharedMap1);
+			compressedIds.set(getIdCompressor(sharedMap3).generateCompressedId(), sharedMap1);
+		}
+
+		for (const [id, map] of compressedIds) {
+			assert(getIdCompressor(map) !== undefined);
+			assert(getIdCompressor(map).normalizeToOpSpace(id) < 0);
+		}
+
+		// Generate DDS ops so that the compressor synchronize
+		sharedMap1.set("key", "value");
+		sharedMap2.set("key2", "value2");
+		sharedMap3.set("key3", "value3");
+
+		await provider.ensureSynchronized();
+
+		for (const [id, map] of compressedIds) {
+			assert(getIdCompressor(map) !== undefined);
+			assert(getIdCompressor(map).normalizeToOpSpace(id) >= 0);
```

Comment:
Similar to above, can you add comments explaining why you expect this value to be >= 0?

Reply:
Could you please add comments to relevant places in all the tests?